docs/src/design/proposals/Registration.md

```diff
@@ -0,0 +1,518 @@
+3. Not handling script loading; the Registration API expects objects,
+   so how the scripts necessary to instantiate those objects (and how they
+   are instantiated) is left to the application developer.
+
+# Evaluation
+
+[Identified problems](#problems-to-address) are addressed by this solution:
+
+1. Dependencies between plugins can be made explicit; a `Plugin` may
+   impose dependencies on other specific `Plugin` subclasses as constructor
+   arguments, disambiguated with JSDoc. The software does not take part
+   in dependency management among plugins; rather, this responsibility is
+   plainly communicated to developers.
+2. Extension points are made explicit; `Provider` instances must be
+   reachable for plugins to configure, and may be made available as
+   public fields of `Plugin`s. Their types can be clearly documented,
+   usages and interactions can be followed with standard developer tools
+   (e.g. breakpoints), and so on.
+3. Reuse of classes between plugins is neither facilitated nor impeded
+   by the registration API. If, however, plugins are written following the
+   "expose classes in namespaces" approach, then it is trivially to
+   expose additional classes in these same namespaces.
+
+This solution offers further benefits:
+
+* The `Provider` API is robust enough to support the various existing
+  extensions of Open MCT, but its usage is opt-in; plugins are free
+  to expose other (potentially wildly different) means of extension.
+  Usage of `Provider`s is _encouraged_ to promote ubiquitous
+  extensibility, but no limitation is exposed.
+* Plugins (and even whole applications, e.g. `MCT`) can be unit tested,
+  in the same style as individual extensions.
+* By moving everything into classes which accept dependencies, a
+  degree of inflexibility is removed from the architecture. In principle,
+  it should be possible to run multiple instances of `MCT` (with
+  their own service instances, etc.) within the same environment. While
+  this is not specifically desirable, it reflects a generally looser
+  coupling between the software and it environment (no expectation of a
+  `bundles.json`, no usage of global state at the language level or
+  effectively-global state at the RequireJS level, etc.) and implies
+  greater flexibility of the application's components.
+* Provides a reduction in code associated with a particular capability;
+  the Framework Layer consists of 18 classes (and has broader
+  responsibilities, including script-loading) whereas this approach
+  consists of 6 classes. (Although some additional work for Angular
+  integration and legacy support will also be needed before achieving
+  parity with the Framework Layer.)
+
+There are some problems with this approach:
+
+* It is highly sensitive to ordering; does not address the problem of
+  [separating configuration from use](http://www.martinfowler.com/articles/injection.html#SeparatingConfigurationFromUse),
```

Comment:
Interesting idea to separate configuration from use.  Would imagine that all configuration is specified per plugin.  e.g.

`MCT.install(new Persistence({type: 'elastic', host: '...'}));`
where you use a plugin which is constructed with some json that then determines which persistence provider to register.

While still allowing direct registration of a provider (that could be part of a plugin), e.g.

`MCT.persistenceService.register(new Persistence.ElasticSearchPersistenceProvider({host: '...'}));`
Where you directly attach your desired persistence provider, which may be part of a Persistence plugin.

The idea here being that MCT will always have the concept of persistence providers, and a plugin is provided that supplies some basic persistence adapters, but that plugin is not required if you would like to use a separate persistence provider (e.g. VISTA).